### PR TITLE
refactor rcnn, test=dygraph

### PR DESCRIPTION
**Pull request by jerrywgz**
Updates:
1. Unify the RCNN transform operators
2. Remove with_background
3. Add from_config and out_shape to simply the config
4. Update the Mask FPN and mAP reaches 39.0/35.5 which is better than detectron2

TODO:
1. Valid the influence of ```+1``` in box operators
2. Fit batch_size > 1 in evaluation
3. Fit Cascade RCNN architectures
4. Remove strip_pad in mask target to speed up training

The models is based on https://github.com/jerrywgz/Paddle/tree/add_aligned_for_roi_align now

### CODE REVIEW

File: dygraph/ppdet/modeling/proposal_generator/rpn_head.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        inputs (dict): ground truth info
+        """
+        prop_gen = self.train_proposal if self.training else self.test_proposal
+        im_shape = inputs['im_shape']
+        batch_size = im_shape.shape[0]
+        rpn_rois_list = [[] for i in range(batch_size)]
+        rpn_prob_list = [[] for i in range(batch_size)]
+        rpn_rois_num_list = [[] for i in range(batch_size)]
+        for rpn_score, rpn_delta, anchor in zip(scores, bbox_deltas, anchors):
+            for i in range(batch_size):
```
Comment (qingqing01) [inline dygraph/ppdet/modeling/proposal_generator/rpn_head.py:133]: add comments for why not  do batch-computing for prop_gen?

File: dygraph/ppdet/modeling/proposal_generator/rpn_head.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                    im_shape=im_shape[i:i + 1])
+                if rpn_rois.shape[0] > 0:
+                    rpn_rois_list[i].append(rpn_rois)
+                    rpn_prob_list[i].append(rpn_rois_prob)
+                    rpn_rois_num_list[i].append(rpn_rois_num)
+
+        rois_collect = []
+        rois_num_collect = []
+        for i in range(batch_size):
+            if len(scores) > 1:
```
Comment (qingqing01) [inline dygraph/ppdet/modeling/proposal_generator/rpn_head.py:147]: 先fpn循环，后batch循环，和上面保持一致

File: dygraph/ppdet/data/source/coco.py [hunk 3]
```diff
@@ -44,16 +44,12 @@ def parse_dataset(self, with_background=True):
...
         records = []
         ct = 0
 
-        # when with_background = True, mapping category to classid, like:
-        #   background:0, first_class:1, second_class:2, ...
-        catid2clsid = dict({
-            catid: i + int(with_background)
-            for i, catid in enumerate(cat_ids)
-        })
+        catid2clsid = dict({catid: i for i, catid in enumerate(cat_ids)})
```
Comment (yghstill) [inline dygraph/ppdet/data/source/coco.py:52]: 去除with_background后：
- Gt2Solov2TargetOp中的解析gt_class需要自动加1，solov2组target时，label=0时是背景。solov2_head中num_classes也需修正。
- SSDHead中num_classes应设置为num_classes + 1，ssd_loss中由于有OHEM，需要加入background:0，所以gt_class所有标签需要+1。

**Review by qingqing01 — COMMENTED** [review]
Need to upload shape_spec.py

File: dygraph/ppdet/modeling/backbones/resnet.py [hunk 4]
```diff
@@ -483,11 +484,13 @@ def __init__(self,
                     lr=1.0,
                     name=_name))
 
-        self.pool = nn.MaxPool2D(kernel_size=3, stride=2, padding=1)
-
         ch_in_list = [64, 256, 512, 1024]
         ch_out_list = [64, 128, 256, 512]
 
+        self.expansion = 4 if depth >= 50 else 1
```
[the remark this answers is not in the document]
Comment (jerrywgz) [reply dygraph/ppdet/modeling/backbones/resnet.py:489]: Done, expansion 会影响r34的输出channel

File: dygraph/ppdet/modeling/backbones/resnet.py [hunk 5]
```diff
@@ -530,17 +541,24 @@ def forward(self, inputs):
 
 @register
 class Res5Head(nn.Layer):
-    def __init__(self, depth=50, feat_in=1024, feat_out=512):
+    def __init__(self, depth=50):
         super(Res5Head, self).__init__()
+        chs = {50: [1024, 512], }
```
[the remark this answers is not in the document]
Comment (jerrywgz) [reply dygraph/ppdet/modeling/backbones/resnet.py]: Done

File: dygraph/ppdet/modeling/heads/mask_head.py [hunk 6]
```diff
@@ -13,195 +13,202 @@
...
+        kwargs = RoIAlign.from_config(cfg, input_shape)
+        roi_pooler.update(kwargs)
+        kwargs = {'input_shape': input_shape}
+        head = create(cfg['head'], **kwargs)
+        return {
+            'roi_extractor': roi_pooler,
+            'head': head,
+        }
+
+    def set_bbox_head(self, bbox_head):
```
[the remark this answers is not in the document]
Comment (jerrywgz) [reply dygraph/ppdet/modeling/heads/mask_head.py]: Done

File: dygraph/ppdet/modeling/layers.py [hunk 7]
```diff
@@ -598,14 +354,15 @@ def __init__(self,
         self.nms_threshold = nms_threshold
         self.normalized = normalized
         self.nms_eta = nms_eta
-        self.background_label = background_label
         self.return_rois_num = return_rois_num
 
-    def __call__(self, bboxes, score):
+    def __call__(self, bboxes, score, background_label=-1):
```
[the remark this answers is not in the document]
Comment (jerrywgz) [reply dygraph/ppdet/modeling/layers.py:359]: 增加了说明，background_label在有的模型是-1；有的模型是类别数

File: dygraph/ppdet/modeling/post_process.py [hunk 8]
```diff
@@ -3,47 +3,114 @@
...
         if self.nms is not None:
             bboxes, score = self.decode(head_out, rois, im_shape, scale_factor)
-            bbox_pred, bbox_num, _ = self.nms(bboxes, score)
+            bbox_pred, bbox_num, _ = self.nms(bboxes, score, self.num_classes)
         else:
             bbox_pred, bbox_num = self.decode(head_out, rois, im_shape,
                                               scale_factor)
         return bbox_pred, bbox_num
 
+    def get_pred(self, bboxes, bbox_num, im_shape, scale_factor):
```
[the remark this answers is not in the document]
Comment (jerrywgz) [reply dygraph/ppdet/modeling/post_process.py:41]: Done

File: dygraph/ppdet/modeling/proposal_generator/anchor_generator.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from ppdet.core.workspace import register
+from .. import ops
+
+
+@register
+class AnchorGenerator(object):
+    __inject__ = ['anchor_generator', 'anchor_target_generator']
```
[the remark this answers is not in the document]
Comment (jerrywgz) [reply dygraph/ppdet/modeling/proposal_generator/anchor_generator.py]: Done

File: dygraph/ppdet/modeling/proposal_generator/rpn_head.py (hunk 1 above)
Comment (jerrywgz) [reply dygraph/ppdet/modeling/proposal_generator/rpn_head.py:135]: Done

File: dygraph/ppdet/modeling/proposal_generator/rpn_head.py (hunk 2 above)
Comment (jerrywgz) [reply dygraph/ppdet/modeling/proposal_generator/rpn_head.py:151]: 这里只需要对batch循环，增加了说明

**Comment by jerrywgz** [comment]
> Need to upload shape_spec.py

Done, thx

File: dygraph/ppdet/data/source/coco.py (hunk 3 above)
Comment (jerrywgz) [reply dygraph/ppdet/data/source/coco.py:52]: 后续pr支持

**Review by heavengate — APPROVED** [review]

File: dygraph/ppdet/modeling/backbones/resnet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nemonameless) [reply dygraph/ppdet/modeling/backbones/resnet.py:489]: expansion之前直接写的是乘4或者1，只有在R34时为1。self.expansion这样单独抽出来写比较标准。

**Review by qingqing01 — APPROVED** [review]
